dbt/include/global_project/macros/materializations/snapshots/snapshot.sql

```diff
@@ -25,16 +25,20 @@
 
   {% set strategy_macro = strategy_dispatch(strategy_name) %}
   {% set strategy = strategy_macro(model, "snapshotted_data", "source_data", config, target_relation_exists) %}
+  {% set dbt_updated_at_data_type = none %}
 
   {% if not target_relation_exists %}
 
       {% set build_sql = build_snapshot_table(strategy, model['compiled_code']) %}
+      {% set dbt_updated_at_data_type = get_updated_at_column_data_type(build_sql) or none %}
```

Comment:
I read this a few times before realizing that all of these changes are only here to support a warning and do not affect `final_sql`. At first glance it looks like these lines are building up to the materialization sql whereas it's sort of a parallel process for a user warning.

Would you consider breaking this out into a separate macro in this file to make it more clear? Or perhaps a separate block of code in this macro? For example, if `snapshot_get_time_data_type` returns `none`, we don't even need to do the rest of this; so it could all live behind an if block with a comment suggesting it's just for a warning.

Reply:
I'm not sure what parts you think should be moved into a macro?  Collecting the value of dbt_updated_at_data_type needs to be collected in the big if/else because there are different sources of sql. The actual warning is only a line in an if after the big if/else.

Reply:
Ah, yeah, I missed that they are different variables. So what about moving the exception into `get_updated_at_column_data_type`? It looks like the exception only depends on the result of that macro and `snapshot_get_time_data_type`, which could be called in that macro instead. In that case, the `get_` prefix may no longer make sense since the return value would no longer be needed.

Reply:
It would be possible to do it that way, but it's my opinion that it would muddy the purpose of the snapshot_get_time_data_type macro. It's quite possible that we might want to call get_snapshot_time_data_type from a different place in the code, where we don't actually want to throw an exception, just get the data type. It has become very difficult to change the signature or purpose of macros once they have been released, so I think it makes more sense to keep the purpose of macros single purpose and limited. 

Additionally -- and this is clearly a matter of opinion/taste -- I prefer at this point to keep "what is happening" inline and readable, with regard to the check that is performed, instead of hiding it away in a macro. Additional changes are going to be coming up in this area and considering the difficulty of changing a macro signature once it's out there, I'd rather leave that for later. It's always possible to insert a new macro that combines "snapshot_get_time_data_type" and throwing the exception, but if we move the exception into the macro, it's more difficult to change later.

Reply:
One way I can imagine refactoring this so the 'detect condition for warning + raise warning' bit is in its own macro while leaving the two helper getter methods alone is: 
1. create a variable for the input to `get_updated_at_column_data_type` (one of build_sql or snapshot_select_sql)
2. pass that into the new macro which would use it as input to `get_updated_at_column_data_type`, and also call `snapshot_get_time_data_type` and determine whether to raise + raise warning
3. call the new macro here: https://github.com/dbt-labs/dbt-adapters/pull/243/files#diff-678f35d7828b12f8a6fb7f70c01683aadf4f3a1141b1b3d4ab89aa356938c2baR77

Reply:
^ I might be missing something here that makes this suggestion not viable, and if it's not I do also agree that this state is more explainable than overloading the `get_updated_at_column_data_type` method.

Reply:
I've refactored this.

Reply:
I like this approach better. I think it's easier to read and makes the happy path a bit more concise.